Data race in mem PubSub

==================
WARNING: DATA RACE
Write at 0x00c0000b2670 by goroutine 7:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:355 +0x0
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.(*PubSub).Close()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub.go:85 +0x244
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.TestPubSub_Close()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub_test.go:143 +0x24b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202

Previous read at 0x00c0000b2670 by goroutine 10:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:158 +0x0
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.(*PubSub).Publish()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub.go:128 +0x16d
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.TestPubSub_Close.func2()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub_test.go:135 +0x1c7

Goroutine 7 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1239 +0x5d7
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1512 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1510 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1418 +0x3b3
  main.main()
      _testmain.go:45 +0x236

Goroutine 10 (running) created at:
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.TestPubSub_Close()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub_test.go:131 +0x230
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Write at 0x00c0000b2610 by goroutine 7:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:355 +0x0
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.(*PubSub).Close()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub.go:91 +0x319
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.TestPubSub_Close()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub_test.go:143 +0x24b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202

Previous read at 0x00c0000b2610 by goroutine 10:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:158 +0x0
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.(*PubSub).Publish()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub.go:129 +0x1b1
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.TestPubSub_Close.func2()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub_test.go:135 +0x1c7

Goroutine 7 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1239 +0x5d7
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1512 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1510 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1418 +0x3b3
  main.main()
      _testmain.go:45 +0x236

Goroutine 10 (running) created at:
  github.com/trustbloc/orb/pkg/activitypub/service/mempubsub.TestPubSub_Close()
      /Users/bob.stasyszyn/go/src/github.com/trustbloc/orb/pkg/activitypub/service/mempubsub/mempubsub_test.go:131 +0x230
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1194 +0x202
==================
 [activitypub_service] 2021/04/19 19:31:19 UTC - mempubsub.(*PubSub).Close -> INFO [service1] ... publisher/subscriber closed
    testing.go:1093: race detected during execution of test
--- FAIL: TestPubSub_Close (0.01s)

=== CONT  
    testing.go:1093: race detected during execution of test

--- a/pkg/activitypub/service/mempubsub/mempubsub.go
+++ b/pkg/activitypub/service/mempubsub/mempubsub.go
@@ -14,6 +14,7 @@ import (
 	"github.com/ThreeDotsLabs/watermill/message"
 	"github.com/trustbloc/edge-core/pkg/log"
 
+	"github.com/trustbloc/orb/pkg/activitypub/service/lifecycle"
 	service "github.com/trustbloc/orb/pkg/activitypub/service/spi"
 )
 
@@ -38,8 +39,8 @@ type Config struct {
 }
 
 // DefaultConfig returns the default configuration.
-func DefaultConfig() *Config {
-	return &Config{
+func DefaultConfig() Config {
+	return Config{
 		Timeout:     defaultTimeout,
 		Concurrency: defaultConcurrency,
 		BufferSize:  defaultBufferSize,
@@ -51,31 +52,61 @@ func DefaultConfig() *Config {
 // the load across a cluster, a persistent message queue (such as RabbitMQ or Kafka) should
 // instead be used.
 type PubSub struct {
-	*Config
+	*lifecycle.Lifecycle
+	Config
 
 	serviceName     string
 	msgChansByTopic map[string][]chan *message.Message
 	mutex           sync.RWMutex
+	publishChan     chan *entry
 	ackChan         chan *message.Message
+	doneChan        chan struct{}
+}
+
+type entry struct {
+	topic    string
+	messages []*message.Message
 }
 
 // New returns a new publisher/subscriber.
-func New(name string, cfg *Config) *PubSub {
+func New(name string, cfg Config) *PubSub {
 	m := &PubSub{
 		Config:          cfg,
 		serviceName:     name,
 		msgChansByTopic: make(map[string][]chan *message.Message),
+		publishChan:     make(chan *entry, cfg.BufferSize),
 		ackChan:         make(chan *message.Message, cfg.Concurrency),
+		doneChan:        make(chan struct{}),
 	}
 
-	go m.listen()
+	m.Lifecycle = lifecycle.New("httpsubscriber-"+name, lifecycle.WithStop(m.stop))
+
+	go m.processMessages()
+	go m.processAcks()
+
+	// Start the service immediately.
+	m.Start()
 
 	return m
 }
 
 // Close closes all resources.
 func (p *PubSub) Close() error {
-	logger.Infof("[%s] Closing publisher/subscriber...", p.serviceName)
+	p.Stop()
+
+	return nil
+}
+
+func (p *PubSub) stop() {
+	logger.Infof("[%s] Stopping publisher/subscriber...", p.serviceName)
+
+	p.doneChan <- struct{}{}
+
+	logger.Debugf("[%s] ... waiting for publisher to stop...", p.serviceName)
+
+	<-p.doneChan
+
+	logger.Debugf("[%s] ... closing subscriber channels...", p.serviceName)
 
 	p.mutex.Lock()
 	defer p.mutex.Unlock()
@@ -90,14 +121,16 @@ func (p *PubSub) Close() error {
 
 	close(p.ackChan)
 
-	logger.Infof("[%s] ... publisher/subscriber closed", p.serviceName)
-
-	return nil
+	logger.Infof("[%s] ... publisher/subscriber stopped.", p.serviceName)
 }
 
 // Subscribe subscribes to a topic and returns the Go channel over which messages
 // are sent. The returned channel will be closed when Close() is called on this struct.
 func (p *PubSub) Subscribe(_ context.Context, topic string) (<-chan *message.Message, error) {
+	if p.State() != service.StateStarted {
+		return nil, service.ErrNotStarted
+	}
+
 	logger.Debugf("[%s] Subscribing to topic [%s]", p.serviceName, topic)
 
 	p.mutex.Lock()
@@ -114,12 +147,47 @@ func (p *PubSub) Subscribe(_ context.Context, topic string) (<-chan *message.Mes
 // immediately after sending the messages to the Go channel(s), although it will
 // block if the concurrency limit (defined by Config.Concurrency) has been reached.
 func (p *PubSub) Publish(topic string, messages ...*message.Message) error {
+	if p.State() != service.StateStarted {
+		return service.ErrNotStarted
+	}
+
+	p.publishChan <- &entry{
+		topic:    topic,
+		messages: messages,
+	}
+
+	return nil
+}
+
+func (p *PubSub) processMessages() {
+	for {
+		select {
+		case entry := <-p.publishChan:
+			p.publish(entry)
+
+		case <-p.doneChan:
+			p.doneChan <- struct{}{}
+
+			logger.Debugf("[%s] ... publisher has stopped", p.serviceName)
+
+			return
+		}
+	}
+}
+
+func (p *PubSub) processAcks() {
+	for msg := range p.ackChan {
+		go p.check(msg)
+	}
+}
+
+func (p *PubSub) publish(entry *entry) {
 	p.mutex.RLock()
-	msgChans := p.msgChansByTopic[topic]
+	msgChans := p.msgChansByTopic[entry.topic]
 	p.mutex.RUnlock()
 
 	for _, msgChan := range msgChans {
-		for _, m := range messages {
+		for _, m := range entry.messages {
 			// Copy the message so that the Ack/Nack is specific to a subscriber
 			msg := m.Copy()
 
@@ -128,14 +196,6 @@ func (p *PubSub) Publish(topic string, messages ...*message.Message) error {
 			msgChan <- msg
 			p.ackChan <- msg
 		}
-	}
-
-	return nil
-}
-
-func (p *PubSub) listen() {
-	for msg := range p.ackChan {
-		go p.check(msg)
 	}
 }
 

--- a/pkg/activitypub/service/mempubsub/mempubsub_test.go
+++ b/pkg/activitypub/service/mempubsub/mempubsub_test.go
@@ -8,6 +8,7 @@ package mempubsub
 
 import (
 	"context"
+	"errors"
 	"sync"
 	"testing"
 	"time"
@@ -19,8 +20,12 @@ import (
 	service "github.com/trustbloc/orb/pkg/activitypub/service/spi"
 )
 
-func TestNew(t *testing.T) {
-	ps := New("service1", DefaultConfig())
+func TestPubSub_Publish(t *testing.T) {
+	cfg := DefaultConfig()
+
+	cfg.Timeout = 100 * time.Millisecond
+
+	ps := New("service1", cfg)
 	require.NotNil(t, ps)
 
 	t.Run("Ack", func(t *testing.T) {
@@ -104,5 +109,125 @@ func TestNew(t *testing.T) {
 		require.Equal(t, msg.UUID, m.UUID)
 	})
 
+	t.Run("Timeout", func(t *testing.T) {
+		msgChan, err := ps.Subscribe(context.Background(), "topic1")
+		require.NoError(t, err)
+
+		undeliverableChan, err := ps.Subscribe(context.Background(), service.UndeliverableTopic)
+		require.NoError(t, err)
+
+		var mutex sync.Mutex
+		receivedMessages := make(map[string]*message.Message)
+		undeliverableMessages := make(map[string]*message.Message)
+
+		go func() {
+			for msg := range msgChan {
+				// Don't Ack/Nack the message. Should timeout and
+				// result in an undeliverable message.
+				mutex.Lock()
+				receivedMessages[msg.UUID] = msg
+				mutex.Unlock()
+			}
+		}()
+
+		go func() {
+			for msg := range undeliverableChan {
+				mutex.Lock()
+				undeliverableMessages[msg.UUID] = msg
+				mutex.Unlock()
+			}
+		}()
+
+		msg := message.NewMessage(watermill.NewUUID(), []byte("payload1"))
+
+		require.NoError(t, ps.Publish("topic1", msg))
+
+		time.Sleep(1000 * time.Millisecond)
+
+		mutex.Lock()
+		m, ok := receivedMessages[msg.UUID]
+		mutex.Unlock()
+
+		require.True(t, ok)
+		require.Equal(t, msg.UUID, m.UUID)
+
+		mutex.Lock()
+		m, ok = undeliverableMessages[msg.UUID]
+		mutex.Unlock()
+
+		require.True(t, ok)
+		require.Equal(t, msg.UUID, m.UUID)
+	})
+
 	require.NoError(t, ps.Close())
+}
+
+func TestPubSub_Error(t *testing.T) {
+	t.Run("Subscribe when closed -> error", func(t *testing.T) {
+		ps := New("service1", DefaultConfig())
+		require.NotNil(t, ps)
+		require.NoError(t, ps.Close())
+
+		msgChan, err := ps.Subscribe(context.Background(), "topic1")
+		require.True(t, errors.Is(err, service.ErrNotStarted))
+		require.Nil(t, msgChan)
+	})
+
+	t.Run("Publish when closed -> error", func(t *testing.T) {
+		ps := New("service1", DefaultConfig())
+		require.NotNil(t, ps)
+		require.NoError(t, ps.Close())
+
+		err := ps.Publish("topic1", message.NewMessage("123", nil))
+		require.True(t, errors.Is(err, service.ErrNotStarted))
+	})
+}
+
+func TestPubSub_Close(t *testing.T) {
+	ps := New("service1", DefaultConfig())
+	require.NotNil(t, ps)
+
+	msgChan, err := ps.Subscribe(context.Background(), "topic1")
+	require.NoError(t, err)
+
+	var mutex sync.Mutex
+
+	receivedMessages := make(map[string]*message.Message)
+
+	go func() {
+		for msg := range msgChan {
+			time.Sleep(5 * time.Millisecond)
+			msg.Ack()
+
+			mutex.Lock()
+			receivedMessages[msg.UUID] = msg
+			mutex.Unlock()
+		}
+	}()
+
+	go func() {
+		for i := 0; i < 200; i++ {
+			msg := message.NewMessage(watermill.NewUUID(), []byte("payload1"))
+
+			if err := ps.Publish("topic1", msg); err != nil {
+				if err == service.ErrNotStarted {
+					return
+				}
+
+				panic(err)
+			}
+
+			time.Sleep(5 * time.Millisecond)
+		}
+	}()
+
+	time.Sleep(50 * time.Millisecond)
+
+	// Close the service while we're still publishing messages to ensure
+	// we don't panic or encounter race conditions.
+	require.NoError(t, ps.Close())
+
+	mutex.Lock()
+	t.Logf("Received %d messages", len(receivedMessages))
+	mutex.Unlock()
 }

